rust/src/index/vector/graph/persisted.rs

```diff
@@ -0,0 +1,314 @@
+            }
+        }
+        let prefetch_size = self.params.prefetch_byte_size / self.vertex_size + 1;
+        let end = std::cmp::min(self.len(), id as usize + prefetch_size);
+        let batch = self
+            .reader
+            .read_range(id as usize..end, &self.vertex_projection)
+            .await?;
+        assert_eq!(batch.num_rows(), end - id as usize);
+        {
+            let mut cache = self.cache.lock().unwrap();
+            let array = as_fixed_size_binary_array(batch.column(0));
+            for (i, vertex_bytes) in array.iter().enumerate() {
+                let vertex = V::from_bytes(vertex_bytes.unwrap())?;
+                cache.insert(id + i as u32, Arc::new(vertex));
+            }
+            Ok(cache.get(&id).unwrap().clone())
+        }
+    }
+
+    /// Get the neighbors of a vertex, specified by its id.
+    pub async fn neighbors(&self, id: u32) -> Result<Arc<UInt32Array>> {
+        {
+            let mut cache = self.neighbors_cache.lock().unwrap();
+            if let Some(neighbors) = cache.get(&id) {
+                return Ok(neighbors.clone());
+            }
+        }
+        let batch = self
+            .reader
+            .read_range(id as usize..(id + 1) as usize, &self.neighbors_projection)
+            .await?;
+        {
+            let mut cache = self.neighbors_cache.lock().unwrap();
+
+            let array = as_list_array(batch.column(0));
+            if array.len() < 1 {
+                return Err(Error::Index("Invalid graph".to_string()));
+            }
+            let value = array.value(0);
+            let nb_array: &UInt32Array = as_primitive_array(value.as_ref());
+            let neighbors = Arc::new(nb_array.clone());
+            cache.insert(id, neighbors.clone());
+            Ok(neighbors)
+        }
+    }
+}
+
+/// Parameters for writing the graph index.
+pub struct WriteGraphParams {
+    pub batch_size: usize,
+}
+
+impl Default for WriteGraphParams {
+    fn default() -> Self {
+        Self { batch_size: 10240 }
+    }
+}
+
+/// Write the graph to a file.
+pub async fn write_graph<V: Vertex>(
+    graph: &GraphBuilder<V>,
+    object_store: &ObjectStore,
+    path: &Path,
+    params: &WriteGraphParams,
+) -> Result<()> {
+    if graph.is_empty() {
+        return Err(Error::Index("Invalid graph".to_string()));
+    }
+    let binary_size = graph.vertex(0).byte_length();
+    let arrow_schema = Arc::new(ArrowSchema::new(vec![
+        Field::new(
+            "vertex",
+            DataType::FixedSizeBinary(binary_size as i32),
+            false,
+        ),
+        Field::new(
+            "neighbors",
+            DataType::List(Box::new(Field::new("item", DataType::UInt32, true))),
+            false,
+        ),
+    ]));
+    let schema = Schema::try_from(arrow_schema.as_ref())?;
+
+    let mut writer = FileWriter::try_new(object_store, path, &schema).await?;
+    for nodes in graph.nodes.as_slice().chunks(params.batch_size) {
```

Comment:
also, are there any other useful graph metadata that can be written separately that's useful without reading the graph?

Reply:
Yes, so there will be other metadata that stored as protobuf index in the index directory as well.

This PR just serialize the graph itself.

likely, there will be some directory structure like:

```
/dataset/_index/1234-5678/index.pb. <- same metadata pb struct 
/dataset/_index/1234-5679/graph.lance. <- graph file
```

In case of HNSW-ish layered graphs, the directory layout will look like
```
/dataset/_index/1234-5678/index.pb. <- same metadata pb struct 
/dataset/_index/1234-5679/layer0.lance.
/dataset/_index/1234-5679/layer1.lance
...
/dataset/_index/1234-5679/layerN.lance.
```